pydantic/_internal/_fields.py

```diff
@@ -22,6 +22,22 @@
     from ..fields import FieldInfo
 
 
+def get_type_hints_infer_globalns(
+    obj: Any,
+    localns: dict[str, Any] | None = None,
+    include_extras: bool = False,
+) -> dict[str, Any]:
+    module_name = getattr(obj, '__module__', None)
+    globalns: dict[str, Any] | None = None
+    if module_name:
+        try:
+            globalns = sys.modules[module_name].__dict__
+        except KeyError:
+            # happens occasionally, see https://github.com/pydantic/pydantic/issues/2363
+            pass
+    return get_type_hints(obj, globalns=globalns, localns=localns, include_extras=include_extras)
```

Comment:
This logic should be reused in various places (anywhere that you know that `obj` would definitely have its module be valid as the globalns for its type hints; this will be the case any time `obj` is a user-defined type, such as a BaseModel subclass not created with `create_model`, typeddict, namedtuple)

Reply:
This is great. I think it's fine here for now, it's where it's being used.